File: apps/desktop/lib/daemon/ws_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

/// Minimal JSON-RPC/WS client: id-correlated request/response plus
/// notification subscriptions. Mirrors packages/cli/src/wsClient.ts —
/// Phase 1 keeps this simple on purpose, no reconnect/retry logic.
class JsonRpcErrorLike {
  const JsonRpcErrorLike({required this.code, required this.message});

  final int code;
  final String message;
}

typedef Listener = void Function(Object? params);

class WsClient {
  WsClient(String url) : _channel = WebSocketChannel.connect(Uri.parse(url)) {
    _openCompleter = Completer<void>();
    _channel.ready
        .then((_) {
          if (!_openCompleter.isCompleted) _openCompleter.complete();
        })
        .catchError((Object err) {
          if (!_openCompleter.isCompleted) {
            _openCompleter.completeError(
              StateError('failed to connect to $url'),
            );
          }
        });
    _subscription = _channel.stream.listen(
      (event) => _handleMessage(event as String),
      onDone: _handleClose,
      onError: (Object _) => _handleClose(),
    );
  }

  final WebSocketChannel _channel;
  late final StreamSubscription<void> _subscription;
  late final Completer<void> _openCompleter;

  int _nextId = 1;
  final Map<int, Completer<Object?>> _pending = {};
  final Map<String, Set<Listener>> _listeners = {};
  bool _closed = false;

  Future<void> connect() => _openCompleter.future;

  Future<T> request<T>(String method, Object? params) {
    if (_closed) {
      return Future.error(StateError('connection is closed'));
    }
    final id = _nextId++;
    final completer = Completer<Object?>();
    _pending[id] = completer;
    _channel.sink.add(
      jsonEncode({'jsonrpc': '2.0', 'id': id, 'method': method, 'params': params}),
    );
    return completer.future.then((value) => value as T);
  }

  /// Subscribes to a notification method; returns an unsubscribe function.
  void Function() on(String method, Listener listener) {
    final set = _listeners.putIfAbsent(method, () => <Listener>{});
    set.add(listener);
    return () => set.remove(listener);
  }

  void close() {
    if (_closed) return;
    _handleClose();
    _subscription.cancel();
    _channel.sink.close();
  }

  void _handleMessage(String raw) {
    Map<String, dynamic> msg;
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map) return;
      msg = Map<String, dynamic>.from(decoded);
    } catch (_) {
      return;
    }

    if (msg.containsKey('id') && msg['id'] != null) {
      final id = msg['id'];
      if (id is! int) return;
      final pending = _pending.remove(id);
      if (pending == null) return;
      if (msg['error'] != null) {
        final error = Map<String, dynamic>.from(msg['error'] as Map);
        pending.completeError(
          StateError(error['message'] as String? ?? 'unknown error'),
        );
      } else {
        pending.complete(msg['result']);
      }
      return;
    }

    final method = msg['method'];
    if (method is String) {
      final set = _listeners[method];
      if (set != null) {
        for (final listener in Set<Listener>.from(set)) {
          listener(msg['params']);
        }
      }
    }
  }

  void _handleClose() {
    _closed = true;
    for (final pending in _pending.values) {
      pending.completeError(StateError('connection closed'));
    }
    _pending.clear();
  }
}

File: packages/cli/src/wsClient.ts
/**
 * Minimal JSON-RPC/WS client: id-correlated request/response plus notification
 * subscriptions. Phase 1 keeps this simple on purpose — no reconnect/retry logic.
 */
export interface JsonRpcErrorLike {
	code: number;
	message: string;
}

type Listener = (params: unknown) => void;

export class WsClient {
	private readonly ws: WebSocket;
	private nextId = 1;
	private readonly pending = new Map<
		number,
		{ resolve: (value: unknown) => void; reject: (err: Error) => void }
	>();
	private readonly listeners = new Map<string, Set<Listener>>();
	private readonly openPromise: Promise<void>;
	private closed = false;

	constructor(url: string) {
		this.ws = new WebSocket(url);
		this.openPromise = new Promise((resolve, reject) => {
			this.ws.onopen = () => resolve();
			this.ws.onerror = () => reject(new Error(`failed to connect to ${url}`));
		});
		this.ws.onmessage = (event) => this.handleMessage(String(event.data));
		this.ws.onclose = () => this.handleClose();
	}

	async connect(): Promise<void> {
		await this.openPromise;
	}

	request<TResult = unknown>(method: string, params: unknown): Promise<TResult> {
		if (this.closed) {
			return Promise.reject(new Error("connection is closed"));
		}
		const id = this.nextId++;
		return new Promise<TResult>((resolve, reject) => {
			this.pending.set(id, {
				resolve: resolve as (value: unknown) => void,
				reject,
			});
			this.ws.send(JSON.stringify({ jsonrpc: "2.0", id, method, params }));
		});
	}

	/** Subscribes to a notification method; returns an unsubscribe function. */
	on(method: string, listener: Listener): () => void {
		let set = this.listeners.get(method);
		if (!set) {
			set = new Set();
			this.listeners.set(method, set);
		}
		set.add(listener);
		return () => set?.delete(listener);
	}

	close(): void {
		this.closed = true;
		this.ws.close();
	}

	private handleMessage(raw: string): void {
		let msg: Record<string, unknown>;
		try {
			msg = JSON.parse(raw);
		} catch {
			return;
		}

		if ("id" in msg && msg.id !== null && msg.id !== undefined) {
			const pending = this.pending.get(msg.id as number);
			if (!pending) return;
			this.pending.delete(msg.id as number);
			if (msg.error) {
				const err = msg.error as JsonRpcErrorLike;
				pending.reject(new Error(err.message));
			} else {
				pending.resolve(msg.result);
			}
			return;
		}

		if (typeof msg.method === "string") {
			const set = this.listeners.get(msg.method);
			if (set) {
				for (const listener of set) listener(msg.params);
			}
		}
	}

	private handleClose(): void {
		this.closed = true;
		for (const [, pending] of this.pending) {
			pending.reject(new Error("connection closed"));
		}
		this.pending.clear();
	}
}
